src/runtime/heap/mem_debug.c

```diff
@@ -0,0 +1,199 @@
+    #endif
+    hdr->sig = DBG_HDR_SIG;
+    hdr->allocsize = b;
+    hdr->padsize = padding;
+    u8 *buf = (u8 *)hdr + padding;
+    #ifdef MEMDBG_INIT
+    set_pattern(buf, b, &pat_init, sizeof(pat_init));
+    #endif
+    return u64_from_pointer(buf);
+}
+
+static void dealloc_check(volatile mem_debug_hdr hdr, u64 a, bytes b, bytes nb, bytes padding)
+{
+    assert(hdr->sig == DBG_HDR_SIG);
+    assert(b == hdr->allocsize);
+    #ifdef MEMDBG_OVERRUN
+    assert(check_pattern(hdr + 1, padding - sizeof(*hdr), &pat_redzone, sizeof(pat_redzone)));
+    assert(check_pattern(pointer_from_u64(a + b), padding, &pat_redzone, sizeof(pat_redzone)));
+    #endif
+    #ifdef MEMDBG_FREE
+    set_pattern(hdr, nb, &pat_freed, sizeof(pat_freed));
+    #endif
+}
+
+static u64 mem_debug_alloc(heap h, bytes b)
+{
+    mem_debug_heap mdh = (mem_debug_heap)h;
+    bytes padding, nb;
+
+    get_debug_alloc_size(b, mdh->padsize, &nb, &padding);
+    mem_debug_hdr hdr = allocate(mdh->parent, nb);
+    if (hdr == INVALID_ADDRESS)
+        return INVALID_PHYSICAL;
+    hdr->alloc_addr = u64_from_pointer(__builtin_extract_return_addr(__builtin_return_address(0)));
+    return alloc_check(hdr, b, padding);
+}
+
+static void mem_debug_dealloc(heap h, u64 a, bytes b)
+{
+    mem_debug_heap mdh = (mem_debug_heap)h;
+    bytes padding, nb;
+
+    get_debug_alloc_size(b, mdh->padsize, &nb, &padding);
+    mem_debug_hdr hdr = (mem_debug_hdr)pointer_from_u64(a - padding);
+    dealloc_check(hdr, a, b, nb, padding);
+    deallocate(mdh->parent, u64_from_pointer(hdr), nb);
+}
+
+heap mem_debug(heap meta, heap parent, u64 padsize)
+{
+    build_assert(PAD_MIN > sizeof(mem_debug_hdr));
+    mem_debug_heap mdh = allocate(meta, sizeof(*mdh));
+    mdh->parent = parent;
+    mdh->h.alloc = mem_debug_alloc;
+    mdh->h.dealloc = mem_debug_dealloc;
+    mdh->padsize = MAX(padsize, PAD_MIN);
+    return &mdh->h;
+}
+
+#ifdef KERNEL
+static void *mem_debug_backed_alloc_map(backed_heap h, bytes b, u64 *phys)
+{
+    mem_debug_backed_heap mdh = (mem_debug_backed_heap)h;
+    bytes padding, nb;
+    u64 nphys;
+
+    get_debug_alloc_size(b, mdh->padsize, &nb, &padding);
+    mem_debug_hdr hdr = alloc_map(mdh->parent, nb, &nphys);
+    if (hdr == INVALID_ADDRESS)
+        return INVALID_ADDRESS;
+    assert(alloc_check(hdr, b, padding) == u64_from_pointer(hdr) + padding);
+    hdr->alloc_addr = u64_from_pointer(__builtin_extract_return_addr(__builtin_return_address(0)));
+    if (phys)
+        *phys = nphys + padding;
+    return (u8 *)hdr + padding;
+}
+
+static void mem_debug_backed_dealloc_unmap(backed_heap h, void *v, u64 p, bytes b)
+{
+    mem_debug_backed_heap mdh = (mem_debug_backed_heap)h;
+    bytes padding, nb;
+    u64 a = u64_from_pointer(v);
+
+    get_debug_alloc_size(b, mdh->padsize, &nb, &padding);
+    mem_debug_hdr hdr = (mem_debug_hdr)pointer_from_u64(a - padding);
+    dealloc_check(hdr, a, b, nb, padding);
+    dealloc_unmap(mdh->parent, hdr, 0, nb);
```

Comment:
I think this should set `p - padding` as physical address argument (if `p` is not zero), so that dealloc_unmap() doesn't have to call physical_from_virtual().

Reply:
Good point, thanks!